### ISSUE TITLE
Add support for relative paths when the MDX node is sourced from a File node

### ISSUE THREAD

**Issue by avigoldman** [ChristopherBiscardi/gatsby-mdx#133]
**Is your feature request related to a problem? Please describe.**
As pointed out in #130, relative imports will fail when pages are programmatically created.

**Describe the solution you'd like**
Relative imports in MDX nodes should just work when the nodes are from a File node.

**Describe alternatives you've considered**
We could ask everyone to add to webpack config to make their imports absolute, but that seems like we are forcing people down path they might not want to take.

**Additional context**
The functionality for this is fairly straightforward. Once #132 and #123 are merged, MDX nodes will have a `fileAbsolutePath` when they are created from File nodes.

Inside the `scope` field creation in the `gatsby/extend-node-type.js`, we can add something similar to the following logic.

```js
// `imports` is an array of import statements
if (mdxNode.fileAbsolutePath) {
  imports = imports.map((importStatement) => {
    const [start, end] = importStatement.split(/\s+from\s+/);
    const importPath = end.replace(/"|'|;/g, '').trim();
    
    if (importPath.startsWith('.')) {
      const absoluteImportPath = path.resolve(path.dirname(mdxNode.fileAbsolutePath), importPath);
      const newImport = `${start} from "${absoluteImportPath}";`
      return newImport;
    }

    return importStatement
  })
}
```

We should probably modify the `babel-plugin-pluck-imports` plugin to export the paths along with the identifiers and full import statements so we can skip regex parsing the import statements.

**Comment by felipeleusin** [comment]
Hi, can I help with this? Was considering moving our documentation system to Gatsby and this seems like a logical choice but no relative import seems like a deal breaker. #123 is open since September, can I help get this rolling?

**Comment by ChristopherBiscardi** [comment]
@felipeleusin I'm happy to have you contribute, but first have you seen that this is already supported if you use the src/pages support? https://github.com/ChristopherBiscardi/gatsby-starter-mdx-basic

Depending on how you're planning to use MDX, the above could be a better solution.

**Comment by felipeleusin** [comment]
Yeah. The problem is we keep our documentation files along our components. So we have /components/Button.tsx and /components/Button.mdx

I'm trying to get this to work without having to change the strcture, going to poke around and see what I find

**Comment by ChristopherBiscardi** [comment]
@felipeleusin The main change needed to support this is in the [babel-plugin-pluck-imports](https://github.com/ChristopherBiscardi/gatsby-mdx/blob/3b90ff236f8ae19ab8c9ecccb4177e0c32834a55/packages/babel-plugin-pluck-imports/index.js) code. 

It's applied to create the [scopeImports](https://github.com/ChristopherBiscardi/gatsby-mdx/blob/3b90ff236f8ae19ab8c9ecccb4177e0c32834a55/packages/gatsby-mdx/utils/gen-mdx.js#L57) when we deal with all of the other MDX processing.

Basically, we write those scope imports out to a file so that webpack includes any dependencies. The change that needs to happen is to turn relative imports into absolute paths so that it doesn't matter where we write the scope file out to.

So the only change that you should need to make is to take the relative path imports and turn them into absolute paths in the babel plugin.

I'm mostly curious about how this change interacts with AST plugins like gatsby-remark-images, but to start don't worry about that and just focus on getting relative imports to work by converting to the absolute path. We can merge the absolute path conversion and worry about gatsby-remark-* support after that.

Happy to answer any questions you have about it [on twitter](https://twitter.com/chrisbiscardi) or here.

**Comment by ChristopherBiscardi** [comment]
Note that an alternative path to writing this babel plugin functionality would be to use the source-filesystem plugin and make your createPage calls point directly to the MDX file

```js
createPage({
  path: '/somewhere',
  component: mdxNode.fileAbsolutePath,
})
```

This works because we have a webpack loader that handles .mdx files, and mdx compiles to a module that exports a default react component, so you can use it directly as a component in createPage.

**Comment by felipeleusin** [comment]
Hmmm, going with this last option was really easy to get it working. Now I'll try to find a way to get the information out of the mdx for things like building a path, but that should be ok to tackle. I'm really new into both gatsby and mdx. Really thanks a lot for the help.

I think my goal will probably be developing something like a plugin gatsby-mdx-local-docs so that's easy to adopt

**Comment by ChristopherBiscardi** [comment]
@felipeleusin You can build a theme to do that. Any logic you end up putting in gatsby-node can go straight into [a theme](https://www.gatsbyjs.org/blog/2018-11-11-introducing-gatsby-themes/).

> Now I'll try to find a way to get the information out of the mdx for things like building a path

You can query the frontmatter or any exports through the same graphql interface you're using to call createPage.

**Comment by ChristopherBiscardi** [comment]
Looks like this is getting solved in #221

**Comment by ChristopherBiscardi** [comment]
Merged in master
